### PR TITLE
[MERGE] base/mail: fix multiple issues in email address handling after #35929

### PR DESCRIPTION
**Pull request by Julien00859**
[FIX] tools: correctly format emails when having an encoding issue
[IMP] tools: restore P2 formataddr default behavior 
[FIX] ir.mail.server: squash redundant CRs (bpo-34424)

Task: 2003936

### CODE REVIEW

**Comment by Julien00859** [comment]
@tde-banana-odoo @odony

File: odoo/tools/mail.py [hunk 1]
```diff
@@ -550,3 +552,114 @@ def formataddr(pair):
                 name=email_addr_escapes_re.sub(r'\\\g<0>', name),
                 addr=address)
     return address
+
+
+# Backport of bpo-34424
+if sys.version_info < (3, 7, 4):
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/mail.py]: xmo suggestion was to adapt the bytecode but I'm pretty sure he was trolling me :thinking:

I tried a workaround involving a regexp to remove those redundant carrial returns for email headers (to, from, ...) but I had troubles concerning the maxlength when the header value is not on the same line as the header key.

Best solution I have in mind is to re-parse the headers and reformat them correctly...

File: odoo/tools/mail.py [hunk 2]
```diff
@@ -516,9 +517,10 @@ def email_escape_char(email_address):
 def decode_message_header(message, header, separator=' '):
     return separator.join(h for h in message.get_all(header, []) if h)
 
-def formataddr(pair):
-    """Takes a 2-tuple of the form (realname, email_address) and returns
-    the string value suitable for an RFC 2822 From, To or Cc header.
+def formataddr(pair, charset='utf-8'):
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/mail.py]: Those people with non-ascii email addresses really..... Let me reach @len-odoo to see what is his opinion

File: odoo/tools/mail.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/mail.py]: If we want to support idna-2008, there is a [pypi](https://pypi.org/project/idna/) module for that (aiohttp uses it)

File: odoo/tools/mail.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/mail.py]: aiohttp uses an external [idna](https://pypi.org/project/idna/) module to handle IDNA-2008, if we want to support the encoding, we should use that library

[the remark this answers is not in the document]
Comment (len-odoo) [reply odoo/tools/mail.py]: Regarding the IDNA encoding, there's an updated version:
https://github.com/odoo/odoo/pull/26756
Which should be straightforward enough to forwardport.
(it uses idna indeed)
(It was refreshed when AL said we should merge it, but then buried when he said it was useless a few days later.)
I think whether it should be merged is your call.

File: odoo/tools/mail.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/mail.py]: After discussion, we opted for `re.sub('\r+', '\r', message_str)`

File: odoo/tools/mail.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/mail.py]: This is going in another pr :)

**Comment by Julien00859** [comment]
All done, thank you for the extensive review :smiley:

**Comment by Julien00859** [comment]
Hello @odony can we merge this one ? :slightly_smiling_face:

**Comment by Julien00859** [comment]
There is a customer on v13 that is affected by the py3 behavior, if it is possible to review this one so I can backport it there

**Comment by odony** [comment]
> There is a customer on v13 that is affected by the py3 behavior, if it is possible to review this one so I can backport it there

Could you elaborate on what you intend to do? This PR is an afterthought above #35929, and it cannot be cherry-picked in 13.0 at all as far as I can see. For instance the old email API did not perform the appropriate header encoding transparently, did it?

**Comment by Julien00859** [comment]
@odony opw-2115767, a customer is having problem in v13 because the python formataddr show the name as quoted printable instead of utf-8 on various form views, this prevent him from migrating from v12 (where the names are showed as utf-8). My plan is to backport c024d894bf (our own formataddr), a66e16e39a (tde's fix) and 81003c1321 (p2 behavior)

**Pull request closed without merging**